// mat-timepicker/esm2020/lib/mat-timepicker.module.mjs
import { MatDialogModule } from '@angular/material/dialog';
import { MatButtonModule } from '@angular/material/button';
import { MatToolbarModule } from '@angular/material/toolbar';
import { MatIconModule } from '@angular/material/icon';
import { MatInputModule } from '@angular/material/input';
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ClockComponent } from './clock/clock.component';
import { MatTimepickerComponentDialogComponent } from './timepicker-dialog/timepicker-dialog.component';
import { MatTimepickerDirective } from './timepicker.directive';
import * as i0 from "@angular/core";
export class MatTimepickerModule {
}
MatTimepickerModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.1.1", ngImport: i0, type: MatTimepickerModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
MatTimepickerModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "14.1.1", ngImport: i0, type: MatTimepickerModule, declarations: [ClockComponent,
        MatTimepickerDirective,
        MatTimepickerComponentDialogComponent], imports: [CommonModule,
        MatDialogModule,
        MatButtonModule,
        MatToolbarModule,
        MatIconModule,
        MatInputModule], exports: [MatTimepickerDirective] });
MatTimepickerModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "14.1.1", ngImport: i0, type: MatTimepickerModule, imports: [CommonModule,
        MatDialogModule,
        MatButtonModule,
        MatToolbarModule,
        MatIconModule,
        MatInputModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.1.1", ngImport: i0, type: MatTimepickerModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [
                        ClockComponent,
                        MatTimepickerDirective,
                        MatTimepickerComponentDialogComponent
                    ],
                    imports: [
                        CommonModule,
                        MatDialogModule,
                        MatButtonModule,
                        MatToolbarModule,
                        MatIconModule,
                        MatInputModule
                    ],
                    exports: [
                        MatTimepickerDirective
                    ]
                }]
        }] });
